gremlin-client/src/structure/value.rs

```diff
@@ -408,3 +409,121 @@ macro_rules! impl_try_from_option {
+{
+    glist
+        .iter()
+        .map(|x| x.clone().try_into())
+        .collect::<GremlinResult<HashSet<T>>>()
+}
+
+fn for_set<T>(gset: &Set) -> GremlinResult<HashSet<T>>
+where
+    T: std::convert::TryFrom<GValue, Error = GremlinError> + Hash + Eq,
+{
+    gset
+        .iter()
+        .map(|x| x.clone().try_into())
+        .collect::<GremlinResult<HashSet<T>>>()
+}
+
+
+macro_rules! impl_try_from_set {
+    ($t:ty) => {
+        impl std::convert::TryFrom<GValue> for HashSet<$t> {
+            type Error = crate::GremlinError;
+
+            fn try_from(value: GValue) -> GremlinResult<Self> {
+                match value {
+                    GValue::List(s) => for_list_to_set(&s),
+                    GValue::Set(s) => for_set(&s),
+                    _ => Err(GremlinError::Cast(format!(
+                        "Cannot cast {:?} to HashSet",
+                        value
+                    ))),
+                }
+            }
+        }
+
+        impl std::convert::TryFrom<&GValue> for HashSet<$t> {
+            type Error = crate::GremlinError;
+
+            fn try_from(value: &GValue) -> GremlinResult<Self> {
+                match value {
+                    GValue::List(s) => for_list_to_set(s),
+                    GValue::Set(s) => for_set(s),
+                    _ => Err(GremlinError::Cast(format!(
+                        "Cannot cast {:?} to HashSet",
+                        value
+                    ))),
+                }
+            }
+        }
+    };
+}
+
+impl_try_from_set!(String);
+// impl_try_from_set!(i32);
+// impl_try_from_set!(i64);
+// impl_try_from_set!(f32);
+// impl_try_from_set!(f64);
+// impl_try_from_set!(Date);
+// impl_try_from_set!(uuid::Uuid);
+// impl_try_from_set!(bool);
+
+macro_rules! impl_try_from_list {
+    ($t:ty) => {
+        impl std::convert::TryFrom<GValue> for Vec<$t> {
+            type Error = crate::GremlinError;
+
+            fn try_from(value: GValue) -> GremlinResult<Self> {
+                match value {
+                    GValue::List(s) => for_list(&s),
+                    GValue::String(s) => Ok(vec![s.into()]),
+                    _ => Err(GremlinError::Cast(format!(
+                        "Cannot cast {:?} to Vec",
+                        value
+                    ))),
+                }
+            }
+        }
+
+        impl std::convert::TryFrom<&GValue> for Vec<$t> {
+            type Error = crate::GremlinError;
+
+            fn try_from(value: &GValue) -> GremlinResult<Self> {
+                match value {
+                    GValue::List(s) => for_list(s),
+                    GValue::String(s) => Ok(vec![s.into()]),
+                    _ => Err(GremlinError::Cast(format!(
+                        "Cannot cast {:?} to Vec",
+                        value
+                    ))),
+                }
+            }
+        }
+    };
+}
+
+impl_try_from_list!(String);
+// impl_try_from_list!(i32);
```

Comment:
Still WIP for these types, going to add them to the tests next, but wanted to show the work so far for feedback